File: backend/models/camera.py
"""Camera-related API models."""

from pydantic import BaseModel, ConfigDict, Field


class CameraInfo(BaseModel):
    """One camera device visible to the local machine."""

    model_config = ConfigDict(extra="forbid")

    index: int = Field(ge=0)
    name: str
    available: bool


class CameraListData(BaseModel):
    """Response payload for camera enumeration."""

    model_config = ConfigDict(extra="forbid")

    cameras: list[CameraInfo]


class CameraStatusData(BaseModel):
    """Current state of the selected camera."""

    model_config = ConfigDict(extra="forbid")

    opened: bool
    index: int | None = None
    name: str | None = None
    available: bool
    width: int | None = None
    height: int | None = None
    fps: float | None = None


class CameraOpenRequest(BaseModel):
    """Camera selection request. The index is never assumed by API clients."""

    model_config = ConfigDict(extra="forbid")

    index: int = Field(default=0, ge=0)


class CaptureData(BaseModel):
    """Metadata returned after a frame is persisted."""

    model_config = ConfigDict(extra="forbid")

    image_path: str
    image_url: str
    captured_at: str

File: frontend/src/types/camera.ts
export type CameraState =
  | "initializing"
  | "available"
  | "not_found"
  | "open_failed"
  | "busy"
  | "disconnected"
  | "read_failed"
  | "mock"
  | "closed";

export type CameraSourceType = "real" | "mock";

export interface CameraInfo {
  index: number;
  name: string;
  available: boolean;
}

export interface CameraListData {
  cameras: CameraInfo[];
  state: CameraState;
  message: string | null;
  code: string | null;
}

export interface CameraStatus {
  state: CameraState;
  source: CameraSourceType | null;
  camera_id: string | null;
  opened: boolean;
  index: number | null;
  name: string | null;
  available: boolean;
  width: number | null;
  height: number | null;
  fps: number | null;
  message: string | null;
  code: string | null;
}

export interface CaptureResult {
  image_path: string;
  image_url: string;
  captured_at: string;
}
